# Day 1, Exercise 1 - understanding data types, variables and basic operations

### Run each command and try to understand why certain operations behave the way they do before looking at the answers.

### There are 4 parts to this exercise, with the answers written under each section
1. Numbers
2. Strings
3. Lists
4. Sets

<hr style="border: 2px solid #000080;">

## 1. Numbers
```python
2 + 2

10 / 4

50 - 5*6    

(50 - 5)*6

5 ** 2

7 % 2

7 // 2

3.14 + 2    
```

- Assign the integer 20 to the variable `width`

- Assign the float 5 to the variable `height`

- Save `width * height` to the variable `area`

- What is the area?

 What happens if you assign 30 to `width` and print `area` again?  

 Re-assign `width * height` to `area` and see if the results changes now.  

- Assign `area * 2` to `2timesArea`

- Assign `area + 2` to `return`


___

### The answers

In [36]:
2 + 2         

4

In [37]:
10 / 4          # Note that division of two integers becomes a float

2.5

In [38]:
50 - 5*6        # Multiplication is evaluated before the minus

20

In [39]:
(50 - 5)*6      # Order of operations, parenthesis first

270

In [40]:
5 ** 2          # Exponentiation

25

In [41]:
7 % 2           # Modulo yields the remainder of the division of 7 by 2 (7 divided by 2 has a remainder of 1)

1

In [42]:
7 // 2          # Floor division returns the largest integer less than or equal to the division result

3

In [43]:
3.14 + 2

5.140000000000001

So why is the example above not exactly 5.14? Python represents floats as base 2 (binary) fractions. This causes some numbers to be approximated rather than exact. If you need very precise calculations, you should be aware that Python behaves this way. To read more about this, click [here](https://docs.python.org/3/tutorial/floatingpoint.html)

In [44]:
width  = 20
height = 5.0
area   = width * height
area                                 # Multiplication of a float and an integer results in a float

100.0

In [45]:
width = 30
area

100.0

When assigning `width * height` to area, python stores the actual value assigned to area in memory, ie `area = 100.0`. It does not store the pointer to `width * height`. To update area with the new inputs we have to actively assign area a value again.

In [46]:
area = width * height
area

150.0

In [47]:
2timesArea = area * 2

SyntaxError: invalid decimal literal (1232134827.py, line 1)

In [ ]:
return = area + 2

Why does the above fail? In the first case we are trying to assign a value to a variable name starting with a number, which is not allowed. In the second case we are trying to assign a value to a reserved keyword. However, both of them results in the same error message, a SyntaxError, which sometimes can be tricky to track, so be aware.

<br><br><br>

<hr style="border: 2px solid #000080;">

## 2. Strings

```python
'Hello World!'
"Hello World!"

'Oh, no, you didn\\'t...'
"Oh, no, you didn't..."

"\\"Yes\\", I did!"
'"Yes", I did'


'Hello'     '  '   'World'     '!'

s = 'Hello\tWorld'
s
print(s)


s = 'First line.\nSecond line.'
s
print(s)

folder = 'C:\some\name'
print(folder)

folder = r'C:\some\name'
print(folder)

'hmm ' + 3 * 'miam'

'Py' 'thon'

prefix = 'Py'
prefix + 'thon'

text = ('Put several strings within parentheses ' +
        'to have them joined together, ' +
        'with the use of ' +
        'concatenation')
text       
```
___

### The answers

In [48]:
'Hello World!'          # "" and '' can be used interchangeably

'Hello World!'

In [49]:
"Hello World!"  

'Hello World!'

In [50]:
'Oh, no, you didn\'t...'

"Oh, no, you didn't..."

In the example above we have to escape the special character `'` to prevent Python from interpreting it as the end of the string. Another way of inserting `'` into a string is to enclose the string in `"` instead, as in the example below. This way, the `'` is interpreted as a normal character.

In [51]:
"Oh, no, you didn't..."

"Oh, no, you didn't..."

In [52]:
"\"Yes\", I did!"

'"Yes", I did!'

In [53]:
'"Yes", I did'

'"Yes", I did'

In [54]:
'Hello'          ' '         'World'             '!'    # Ignores spaces between strings

'Hello World!'

In [55]:
s = 'Hello\tWorld'                                      # Adds a tab between the words
s                                                       # Notice that s only shows the content

'Hello\tWorld'

In [56]:
print(s)                                                # While print() renders the content of s

Hello	World


In [57]:
s = 'First line.\nSecond line.'                         # \n adds a newline
s                                                       # Without print(), \n is included in the output

'First line.\nSecond line.'

In [58]:
print(s)                                                # With print(), \n produces a new line

First line.
Second line.


In [59]:
folder = 'C:\some\name'                                 # Here the \n will be interpreted as a newline
print(folder)

C:\some
ame


In [60]:
folder = r'C:\some\name'
print(folder)

C:\some\name


If you don’t want characters prefaced by `\` to be interpreted as special characters, you can use __raw strings__ by adding an `r` before the first quote as above

In [61]:
'hmm ' + 3 * 'miam'

'hmm miammiammiam'

In [62]:
'Py' 'thon'

'Python'

In [63]:
prefix = 'Py'
prefix + 'thon'

'Python'

In [64]:
text = ('Put several strings within parentheses ' +
        'to have them joined together, ' +
        'with the use of ' +
        'concatenation')
text

'Put several strings within parentheses to have them joined together, with the use of concatenation'

<hr style="border: 2px solid #000080;">

## 3. Lists

```python
squares = [1, 4, 9, 16, 25] 
cubes   = [1, 8, 27, 64, 125]

squares + cubes 
squares * 2
squares * cubes
```
___

### The answers

Lists are a collection of items grouped together. A list is denoted using comma-separated values between square brackets. A list might contain items of different types, but usually the items all have the same type.

In [65]:
squares = [1, 4, 9, 16, 25] 
cubes = [1, 8, 27, 64, 125]

squares + cubes     # Even though both contains integers and are of equal length they cannot be added together, only concatenated

[1, 4, 9, 16, 25, 1, 8, 27, 64, 125]

In [66]:
squares * 2         # Multiplication with 2 will produce a new list with a duplicate of the first list concatenated to the end

[1, 4, 9, 16, 25, 1, 4, 9, 16, 25]

In [67]:
squares * cubes     # Multiplication of lists can only be done with integers, not lists

TypeError: can't multiply sequence by non-int of type 'list'

<hr style="border: 2px solid #000080;">

## 4. Set

```python
mylist = ["1", "1", "2", "2", "4", "4", "10", "11", "13"]
myset = {"1", "1", "2", "2", "4", "4", "10", "11", "13"}
unique_list = list(set(mylist))
sorted(unique_list)

"10" in myset
"10" in unique_list

set1 = {1, 2, 3}
set2 = {3, 4, 5}
union_set = set1.union(set2)
print(union_set)
```
___

<hr style="border: 2px solid #000080;">

### The answers

A set is an unordered collection of unique elements. It is defined using curly braces `{}` or the `set()` constructor, and they support various operations like union, intersection, and difference.

In [69]:
mylist = ["1", "1", "2", "2", "4", "4", "10", "11", "13"] # Lists may contain duplicate elements
myset = {"1", "1", "2", "2", "4", "4", "10", "11", "13"} # Sets automatically remove duplicate elements. 
unique_list = list(set(mylist)) # a convenient way to remove duplicate elements from a list 
sorted_unique_list = sorted(unique_list) # Strings are sorted in alphabetical order.
print("mylist: ", mylist)
print("myset: ", myset)
print("unique_list: ", unique_list) 
print("sorted_unique_list: ", sorted_unique_list) 


mylist:  ['1', '1', '2', '2', '4', '4', '10', '11', '13']
myset:  {'11', '1', '2', '13', '4', '10'}
unique_list:  ['4', '10', '2', '13', '11', '1']
sorted_unique_list:  ['1', '10', '11', '13', '2', '4']


A list may contain duplicate elements, but a set contains only unique elements. Sets automatically remove duplicate elements when they are assigned with values. 

In [70]:
print("10" in myset)
print("10" in unique_list)

True
True


One can check membership of an element using the `in` operation for both Lists and Sets. However, Sets are usually faster for membership checking, especially when the data contains many elements.

In [71]:
set1 = {1, 2, 3}
set2 = {3, 4, 5}
union_set = set1.union(set2)
print(union_set) # The union method returns a new set with the union of two sets.

{1, 2, 3, 4, 5}
